File: Dsa-day165.py
class Solution:
    def pyramidTransition(self, bottom: str, allowed: List[str]) -> bool:
        trans = defaultdict(list)
        for s in allowed:
            trans[s[:2]].append(s[2])
        
        memo = set()
        
        def dfs(row: str) -> bool:
            if len(row) == 1:
                return True
            if row in memo:
                return False
            
            next_rows = []
            def build(idx: int, path: list):
                if idx == len(row) - 1:
                    next_rows.append(''.join(path))
                    return
                key = row[idx:idx+2]
                if key not in trans:
                    return
                for c in trans[key]:
                    path.append(c)
                    build(idx + 1, path)
                    path.pop()
            
            build(0, [])
            
            for nxt in next_rows:
                if dfs(nxt):
                    return True
            
            memo.add(row)
            return False
        
        return dfs(bottom)

File: Dsa-day165.cpp
class Solution {
public:
    unordered_map<string, vector<char>> trans;
    unordered_set<string> fail;

    bool pyramidTransition(string bottom, vector<string>& allowed) {
        for (const string& s : allowed) {
            string key = s.substr(0, 2);
            trans[key].push_back(s[2]);
        }
        return dfs(bottom);
    }

private:
    bool dfs(const string& row) {
        if (row.size() == 1) return true;
        if (fail.count(row)) return false;

        vector<string> nextRows;
        string curr;
        buildNextRows(row, 0, curr, nextRows);

        for (const string& next : nextRows) {
            if (dfs(next)) return true;
        }

        fail.insert(row);
        return false;
    }

    void buildNextRows(const string& row, int idx, string& curr, vector<string>& res) {
        if (idx == row.size() - 1) {
            res.push_back(curr);
            return;
        }

        string key = row.substr(idx, 2);
        auto it = trans.find(key);
        if (it == trans.end()) return;

        for (char c : it->second) {
            curr.push_back(c);
            buildNextRows(row, idx + 1, curr, res);
            curr.pop_back();
        }
    }
};